benchexec/tablegenerator/react-table/src/utils/utils.js

```diff
@@ -192,6 +192,355 @@ const setHashSearch = (
   document.location.href = hrefString;
 };
 
+const makeUrlFilterDeserializer = (statusValues, categoryValues) => {
+  const deserializer = makeFilterDeserializer({ categoryValues, statusValues });
+  return (str) => {
+    const params = getHashSearch(str);
+    if (params.filter) {
+      const out = deserializer(params.filter);
+      return out;
+    }
+    return null;
+  };
+};
+
+const makeSerializedFilterValue = (filter) => {
+  const parts = [];
+  for (const [key, values] of Object.entries(filter)) {
+    parts.push(`${key}(${values.map(escape).join(",")})`);
+  }
+  return parts.join(",");
+};
+const createDistinctValueFilters = (selected, nominal, trim = false) => {
+  const filter = {};
+  // we want to minimize the needed space to encode the filter
+  // if we have more than half of all values selected, we encode all not selected
+  // values in "notIn", otherwise we encode all selected values in "in"
+  if (selected.length > Math.floor(nominal.length / 2.0)) {
+    const exclusions = [];
+    for (const status of nominal) {
+      if (!selected.includes(status)) {
+        exclusions.push(trim ? status.trim() : status);
+      }
+    }
+    filter.notIn = exclusions;
+  } else {
+    filter.in = selected.map((val) => (trim ? val.trim() : val));
+  }
+  return makeSerializedFilterValue(filter);
+};
+
+const makeFilterSerializer = ({
+  statusValues: allStatusValues,
+  categoryValues: allCategoryValues,
+}) => (filter) => {
+  const groupedFilters = {};
+  for (const { id, value, values } of filter) {
+    if (id === "id") {
+      if (values && values.length > 0) {
+        groupedFilters.ids = { values: values.map((val) => (val ? val : "")) };
+      }
+      continue;
+    }
+    const [tool, name, column] = id.split("_");
+    const toolBucket = groupedFilters[tool] || {};
+    const columnBucket = toolBucket[column] || { name: escape(name) };
+
+    if (allStatusValues[tool][column] || allCategoryValues[tool][column]) {
+      // we are processing a status column with checkboxes
+      if (value.endsWith(" ")) {
+        // category value
+        const selectedCategoryValues = columnBucket.categoryValues || [];
+        selectedCategoryValues.push(value);
+        columnBucket.categoryValues = selectedCategoryValues;
+      } else {
+        // status value
+        const selectedStatusValues = columnBucket.statusValues || [];
+        selectedStatusValues.push(value);
+        columnBucket.statusValues = selectedStatusValues;
```

Comment:
I think you are making a good point here. 
The initial reason why i went for the current approach is because i wanted to make the column type (value vs id vs status) identifiable by its defined properties. Just like a couple of lines down in the else block we assign a value to the `value` property and thereby define it. 
In OOP we would properly use different implementations of a interface or abstract structure that we could then either type check in the code or invoke a type-specific member function.

Having a `statusValues` and `categoryValues` property assigned to every filter might, at least in my immediate opinion, be a little bit misleading as there might be filters without a `value` or `values` property for example. Also, we wouldn't be able to identify a status filter if all status or category values have been deselected (even though this is only an edge-case).

I do however agree that this might not be the cleanest or most obvious solution. What do you think about introducing a `type` property that we could check for in the serialization block below? That way we wouldn't need to make assumptions over the type of a filter depending on if properties exist or if they hold values but by comparing enums.

Reply:
Again I agree with both of you :-)

Reply:
> Having a `statusValues` and `categoryValues` property assigned to every filter might, at least in my immediate opinion, be a little bit misleading as there might be filters without a `value` or `values` property for example. Also, we wouldn't be able to identify a status filter if all status or category values have been deselected (even though this is only an edge-case).
> 
I see your point. I agree that if it's possible that some filters don't have those properties, defining them for every filter wouldn't be a good implementation.
> I do however agree that this might not be the cleanest or most obvious solution. What do you think about introducing a `type` property that we could check for in the serialization block below? That way we wouldn't need to make assumptions over the type of a filter depending on if properties exist or if they hold values but by comparing enums.
>
This would also make it possible to use a single attribute like `columnBucket.value` to store the values instead of using three different ones, right? This sounds like a good solution in my opinion.